### DISCUSSION TITLE
LAMA0041: Exception thrown: The parameter 'uri' of template is not compile-time

### DISCUSSION THREAD

**Discussion by WhitWaldo** [metalama/Metalama#343] [Q&A] [answered]
I've got a solution, available [here](https://github.com/WhitWaldo/Innovian.Aspects.ProxyGenerator), that just isn't working and I'm at a loss as to why.

There are three projects in the solution:
- Innovian.Aspects.ProxyGenerator: Contains the aspects themselves and the various types they've expecting
- Innovian.Aspects.ProxyGenerator.Tests: Intended to contain the various aspect tests
- TestProject: A project from which I've been playing around with the shape of the aspects. No real need to bother with this one.

I've got another aspect in this solution, `ProxyGeneratorAttribute`. Pay this one no heed as it's working fine for my purposes. The aspect I'm having trouble with is `ApiMethodAttribute`.

The idea is that I should be able to indicate an HTTP verb (e.g. GET, POST, DELETE, etc.) and provide a URI path not unlike the template URL for ASP.NET Core API controllers. The aspect should modify the method it's attached to, iterate through the template URL and replace the keys in curly braces with the string values of the parameters with matching case-insensitive names passed into the method itself. A future iteration of this would ideally add an eligibility check that fails if any of the parameter values is not a `string`, doesn't inherit from `string` or doesn't explicitly declare an override for `ToString()`, but that's not in this version.

An earlier iteration of this that I was working on this morning had the arguments for the `GetMethodTemplate` template marked with `[CompileTime]` but this turned into a game of whack-a-mole where I'd get perpetual warnings about trying to use run-time values in compile-time values. I finally got them all to go away simply by clearing those `[CompileTime]` attributes and you'll notice that the `Innovian.Aspects.ProxyGenerator` project builds without error.

Unfortunately, the same cannot be said for the `Innovian.Aspects.ProxyGenerator.Tests` project. When one attempts to build it, the following exception is observed:
> LAMA0041	'Exception of type 'Metalama.Framework.Aspects.InvalidTemplateSignatureException' thrown while executing BuildAspect for aspect [ApiMethodAttribute] applied to 'SampleService.ListAllIdsAsync(string)': The parameter 'uri' of template 'ApiMethodAttribute.GetMethodTemplate(string, IParameterList)' is not compile-time.

The `uri` parameter is passed into the method template via the `args` argument in the `IntroduceMethod` and there, is being sourced from the property on the attribute itself provided by the user decorating the method with the aspect/attribute with their values. So while they're presumably there at run-time, they're being passed into the aspect and thus should be available at compile-time despite the text of the error.

I appreciate any thoughts around how to remedy this - thanks!

**Comment by gfraiteur** [comment]
I would need to see a few pairs of "source code" -> "transformed code" examples to help understand what the aspect should do.

**Comment by gfraiteur** [comment, marked as answer]
I think I may have understood your ApiMethod requirements and here is my implementation:

https://github.com/gfraiteur/Innovian.Aspects.ProxyGenerator/blob/master/Innovian.Aspects.ProxyGenerator/ApiMethodAttribute.cs

**Comment by WhitWaldo** [comment]
First of all, thank you for giving this a look. Having read through it, I've definitely learned several better practices:
- I didn't realize there was a method for skipping aspect application
- The diagnostics are a great way to explain aspect failures and why
- It definitely simplifies the template to shuffle the parameter processing to the `BuildAspect` method and just pass in the results - this makes a lot of sense too, since it just leaves minimal runtime code in the output
- It didn't even occur to me that this would be a great use of the `InterpolatedStringBuilder`

This is working on my side, but I have a lingering question for you about unit testing it. I'd like to hammer out some unit tests for the UrlTemplateTokenizer, but since it's marked with `[CompileTime]`, I cannot reference it with the run-time unit tests. Do you have a recommendation of how to approach this?

**Reply by svick** [reply]
Since `UrlTemplateTokenizer` doesn't use any Metalama types, I think you should be able to mark it as `[RunTimeOrCompileTime]`, so that it can be tested.

**Reply by WhitWaldo** [reply]
Unfortunately, I tried that. After setting the updated attribute to `[RunTimeOrCompileTime]`, the aspect starts giving me a wall of errors about the TokenKind type now being run-time only instead of compile-time.

![image](https://github.com/user-attachments/assets/2b8de17e-81bd-4c90-b7f9-f08fcd37fcad)

**Reply by svick** [reply]
It seems we have a bug with having an unmarked `enum` nested inside a `[RunTimeOrCompileTime]` type. If you mark the `enum` as `[RunTimeOrCompileTime]` as well, it should work.

**Reply by WhitWaldo** [reply]
That fixed it - thank you!

**Reply by gfraiteur** [reply]
Note that the recommended way to unit test compile-time is _not_ how @svick suggested. It is described [here](https://doc.postsharp.net/metalama/conceptual/aspects/testing/compile-time-testing).
